lib/services/local/saml_idp_service_provider.go

```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
 	sp.SetEntityDescriptor(string(entityDescriptor))
 	return nil
 }
+
+// embedAttributeMapping embeds attribute mapping details into entity descriptor.
+func (s *SAMLIdPServiceProviderService) embedAttributeMapping(sp types.SAMLIdPServiceProvider) error {
+	if len(sp.GetAttributeMapping()) == 0 {
+		s.log.Infof("no custom attribute mapping values provided for %s. SAML assertion will default to uid and eduPersonAffiliate", sp.GetEntityID())
+		return nil
+	}
+
+	ed, err := samlsp.ParseMetadata([]byte(sp.GetEntityDescriptor()))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// return if provided attributes are already embedded.
+	embeddedSPSSODescriptorIndex, embeddedAttributes := samlSPSSODescriptorToProtoSAMLAttributeMapping(ed.SPSSODescriptors)
+	if reflect.DeepEqual(sp.GetAttributeMapping(), embeddedAttributes) {
+		return nil
+	}
+
+	newEmbeddedSPSSODescriptor := spSSODescriptorWithRequestedAttributes(sp.GetAttributeMapping())
+	// If there is an existing embedded SPSSODescriptor, we'll replace it with newEmbeddedSPSSODescriptor
```

Comment:
Nit: I would rephrase the comment a bit to clarify that if there's an existing descriptor for "teleport" service, we update and replace it, otherwise we create a new one for "teleport" service and append it. It took me a bit to figure this out.

Reply:
updated in https://github.com/gravitational/teleport/pull/35705/commits/806e5625965fcc580d7c851e2df7f25fd042d7d5. let me know how it reads now.